### ISSUE TITLE
Unable to find the dataset 

### ISSUE THREAD

**Issue by HarshaSatyavardhan** [LiaoShiqi97/DSCM_fundus#4]
I tried to reproduce the results but i am receiving this error

```
  File "/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py", line 444, in fit
    self.data_connector.prepare_data(model)
  File "/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/connectors/data_connector.py", line 60, in prepare_data
    model.prepare_data()
  File "/content/DSCM_fundus/deepscm/experiments/medical/base_experiment.py", line 172, in prepare_data
    data_dir = self.hparams.data_train_dir if hasattr(self.hparams, 'data_dir') else '/mnt/alpha/diabetes/MS/data/images_original/'
  File "/usr/local/lib/python3.9/dist-packages/pytorch_lightning/utilities/parsing.py", line 185, in __getattr__
    raise AttributeError(f'Missing attribute "{key}"') from exp
AttributeError: Missing attribute "data_train_dir"
```
can you elaborate on the datasets you are using for training and from where you are downloading them or generating them

**Comment by LiaoShiqi97** [comment]
Hi, HarshaSatyavardhan.

Thanks for you interest in my work.
The dataset is from  **maastricht study**(https://www.demaastrichtstudie.nl/research/data-guidelines, you can follow this link to apply for dataset)[1]. And there is some alternative dataset you might use as well, **rfmid**(https://ieee-dataport.org/open-access/retinal-fundus-multi-disease-image-dataset-rfmid)
Hope this answers your question.


[1] Miranda T Schram, Simone JS Sep, Carla J van der Kallen, Pieter C Dagnelie, Annemarie
Koster, Nicolaas Schaper, Ronald Henry, and Coen DA Stehouwer. The maastricht study:
an extensive phenotyping study on determinants of type 2 diabetes, its complications and its
comorbidities. European journal of epidemiology, 29(6):439–451, 2014. 31

**Comment by HarshaSatyavardhan** [comment]
Thanks, hey I was trying to use this on some custom dataset other than medical one like CelebA dataset. it's becoming very hard to follow the code base as I am new to both pyro and normalising flows. can you provide some information on how should I need to proceed.

**Comment by LiaoShiqi97** [comment]
Hi HarshaSatyavardhan,
Glad you contact me again. 
Firstly, I suggest you follow the paper Deep Structural Causal Models for Tractable Counterfactual Inference(https://arxiv.org/abs/2006.06485) and its code. It has accesible dataset with itself so you can play with it.
Secondly, I recoomend three documents：1) the offical document (https://docs.pyro.ai/en/stable/distributions.html), 2) a demo for probabilistic programming(https://willcrichton.net/notes/probabilistic-programming-under-the-hood/), 3) an easy guide (https://willcrichton.net/notes/probabilistic-programming-under-the-hood/).

Hope these answer your question.

**Comment by HarshaSatyavardhan** [comment]
Thank you so much for replying fast.
As you Said I have referred to the paper and the blog post. I will try to explain what I understood here please correct me if I am wrong.

so according to the paper what I understood is we are forming a DAG by domain knowledge where data is available with us. Here its sex, age, T2D, image (These are all endogenous variables).

Here in this code in the `BaseCovariateExperiment` class 

https://github.com/LiaoShiqi97/DSCM_fundus/blob/c5896451c57bbb6327ab7629b4f348eb69594051/deepscm/experiments/medical/base_experiment.py#L167-L192

I see loading the fundus images and all the remaining attributes to a 2 dimensional vector but why calculating mean and std for only age ? 
and I see the z getting assigned here and I assume its the prior for every exogenous variable

and in `BaseSEM` class 

https://github.com/LiaoShiqi97/DSCM_fundus/blob/c5896451c57bbb6327ab7629b4f348eb69594051/deepscm/experiments/medical/base_experiment.py#L36-L55

you have a `_get_preprocess_transforms` which you are calling in the BaseVISEM class and inv it there I dont know why you are doing it 

https://github.com/LiaoShiqi97/DSCM_fundus/blob/c5896451c57bbb6327ab7629b4f348eb69594051/deepscm/experiments/medical/fundus/sem_vi/base_sem_experiment.py#L181-L188

Now we infer exogenous variable (latent space or noise) from the endogenous variable by passing them through flow based model.

and you are doing this using infer_exogenous function in the `BaseSEM` class

https://github.com/LiaoShiqi97/DSCM_fundus/blob/c5896451c57bbb6327ab7629b4f348eb69594051/deepscm/experiments/medical/base_experiment.py#L104-L120

which is called in the `BaseVISEM` again 

https://github.com/LiaoShiqi97/DSCM_fundus/blob/c5896451c57bbb6327ab7629b4f348eb69594051/deepscm/experiments/medical/fundus/sem_vi/base_sem_experiment.py#L218-L227

Now using SCM we can change these endogenous variables and recalculate the exogenous for the changed variables (abduction process) to get counterfactuals.

https://github.com/LiaoShiqi97/DSCM_fundus/blob/c5896451c57bbb6327ab7629b4f348eb69594051/deepscm/experiments/medical/fundus/sem_vi/base_sem_experiment.py#L243-L263

what I am confused is the directions I dont see them explicitly in the code base. but I assume that is the order we need to follow when passed into the flow model to get the image 

When I checked the `BaseVISEM` and `ConditionalVISEM`

you have a flow where you apply transformations to `age` into it

https://github.com/LiaoShiqi97/DSCM_fundus/blob/c5896451c57bbb6327ab7629b4f348eb69594051/deepscm/experiments/medical/fundus/sem_vi/base_sem_experiment.py#L166-L171

and flow for `T2D`

https://github.com/LiaoShiqi97/DSCM_fundus/blob/c5896451c57bbb6327ab7629b4f348eb69594051/deepscm/experiments/medical/fundus/sem_vi/conditional_sem.py#L17-L77

I mean according to the DAG directions I don't see data is getting pass through flow models and getting image.
If the directions are not considered then how recalculating the exogenous variables is done and getting a counterfactual image.

https://github.com/LiaoShiqi97/DSCM_fundus/blob/c5896451c57bbb6327ab7629b4f348eb69594051/deepscm/experiments/medical/fundus/sem_vi/base_sem_experiment.py#L406-L417

you are forming batch of image and labels to do VI but what about the directions I assume we have `z` separate for each endogenous variable not a same `z` for every endogenous variable 

am I missing something here?

The post must be long but I am working on this project alone and finding very hard to understand myself, it would be so helpful if you can answer my questions.
